go/kbtest/kbtest.go

```diff
@@ -70,3 +87,128 @@ func CreateAndSignupFakeUser(prefix string, g *libkb.GlobalContext) (*FakeUser,
+	if err != nil {
+		return err
+	}
+	if !prov {
+		return libkb.LoginRequiredError{}
+	}
+	return nil
+}
+
+type testProvisionUI struct {
+	secretCh               chan kex2.Secret
+	method                 keybase1.ProvisionMethod
+	gpgMethod              keybase1.GPGMethod
+	chooseDevice           string
+	verbose                bool
+	calledChooseDeviceType int
+	abortSwitchToGPGSign   bool
+}
+
+func newTestProvisionUI() *testProvisionUI {
+	ui := &testProvisionUI{method: keybase1.ProvisionMethod_DEVICE}
+	if len(os.Getenv("KB_TEST_VERBOSE")) > 0 {
+		ui.verbose = true
+	}
+	ui.gpgMethod = keybase1.GPGMethod_GPG_IMPORT
+	return ui
+}
+
+func (u *testProvisionUI) printf(format string, a ...interface{}) {
+	if !u.verbose {
+		return
+	}
+	fmt.Printf("testProvisionUI: "+format+"\n", a...)
+}
+
+func (u *testProvisionUI) ChooseProvisioningMethod(_ context.Context, _ keybase1.ChooseProvisioningMethodArg) (keybase1.ProvisionMethod, error) {
+	panic("ChooseProvisioningMethod deprecated")
+}
+
+func (u *testProvisionUI) ChooseGPGMethod(_ context.Context, _ keybase1.ChooseGPGMethodArg) (keybase1.GPGMethod, error) {
+	u.printf("ChooseGPGMethod")
+	return u.gpgMethod, nil
+}
+
+func (u *testProvisionUI) SwitchToGPGSignOK(ctx context.Context, arg keybase1.SwitchToGPGSignOKArg) (bool, error) {
+	if u.abortSwitchToGPGSign {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (u *testProvisionUI) ChooseDevice(_ context.Context, arg keybase1.ChooseDeviceArg) (keybase1.DeviceID, error) {
+	u.printf("ChooseDevice")
+	if len(arg.Devices) == 0 {
+		return "", nil
+	}
+
+	if u.chooseDevice == "none" {
+		return "", nil
+	}
+
+	if len(u.chooseDevice) > 0 {
+		for _, d := range arg.Devices {
+			if d.Type == u.chooseDevice {
+				return d.DeviceID, nil
+			}
+		}
+	}
+	return "", nil
+}
+
+func (u *testProvisionUI) ChooseDeviceType(_ context.Context, _ keybase1.ChooseDeviceTypeArg) (keybase1.DeviceType, error) {
+	u.printf("ChooseDeviceType")
+	u.calledChooseDeviceType++
+	return keybase1.DeviceType_DESKTOP, nil
+}
+
+func (u *testProvisionUI) DisplayAndPromptSecret(_ context.Context, arg keybase1.DisplayAndPromptSecretArg) (keybase1.SecretResponse, error) {
+	u.printf("DisplayAndPromptSecret")
+	var ks kex2.Secret
+	copy(ks[:], arg.Secret)
+	u.secretCh <- ks
+	var sr keybase1.SecretResponse
+	return sr, nil
+}
+
+func (u *testProvisionUI) PromptNewDeviceName(_ context.Context, arg keybase1.PromptNewDeviceNameArg) (string, error) {
+	u.printf("PromptNewDeviceName")
+	return libkb.RandString("device", 5)
+}
+
+func (u *testProvisionUI) DisplaySecretExchanged(_ context.Context, _ int) error {
+	u.printf("DisplaySecretExchanged")
+	return nil
+}
+
+func (u *testProvisionUI) ProvisioneeSuccess(_ context.Context, _ keybase1.ProvisioneeSuccessArg) error {
+	u.printf("ProvisioneeSuccess")
+	return nil
+}
+
+func (u *testProvisionUI) ProvisionerSuccess(_ context.Context, _ keybase1.ProvisionerSuccessArg) error {
+	u.printf("ProvisionerSuccess")
+	return nil
+}
```

Comment:
I hesitate to change anything with provisioning (and its tests).

If you'd like to reuse the provision ui in the engine test, you could just make the test there?